rubyforgood/diaper
# [1166] Enable destruction of Donations for organization admin

Resolves #1166

### Description

Organization Admin can destroy a donation from the `Donation#show` view.

I'm unfamiliar with the design choices of this app, so I was not sure where to put the delete button.  See screenshot below.  Let me know if it should be moved.

- The action to destroy these donations already existed, it just wasn't accessible from the UI (as far as I could tell).
- Before this commit, any user who had authorization to access the `#show` action in the `DonationController` was authorized to perform the ``#destroy` action.  I leveraged an existing `authorize_admin` method found in other controllers to enforce authentication.
- Moved `#authorize_admin` method into ApplicationController.  This method was used in two places, this commit introduces a third use.  I felt it was time to move it into the super class.
- The diff for `donation_system_spec.rb` got mangled because I wrapped a lot of tests in a context.  I didn't change _that much_ about those specs.

### Type of change

* New feature (non-breaking change which adds functionality)

### How Has This Been Tested?

As an Organization Admin:

- Visit the donation index for an organization that has at least one donation.
- Note the total quantity of items at the bottom of the table.
- View a donation, note the quantity of items associated with that donation.
- Destroy the donation.
- Confirm that the quantity of items at the bottom of the table has been updated to be equal to the original quantity of items less the quantity of items that were associated with the destroyed donation.  Also confirm that the donation is on longer visible on the table.

As a user who is _not_ an Organization Admin:

- Visit the donation index for an organization that has at least one donation.
- View a donation.
- Confirm you do not see the link/button to destroy the donation.

### Screenshots
The new delete button/link: 

![image](https://user-images.githubusercontent.com/1395957/65543829-99bac580-ded7-11e9-8847-7a1072b9b78d.png)

The same view when the new delete link/button is not enabled:

![image](https://user-images.githubusercontent.com/1395957/65544329-9aa02700-ded8-11e9-82fb-79681037ef9f.png)
